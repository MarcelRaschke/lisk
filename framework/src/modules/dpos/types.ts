/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { BlockHeader } from '@liskhq/lisk-chain';

export interface RegisteredDelegate {
	readonly username: string;
	readonly address: Buffer;
}

export interface RegisteredDelegates {
	registeredDelegates: RegisteredDelegate[];
}

export interface DelegatePersistedUsernames {
	readonly registeredDelegates: RegisteredDelegate[];
}

export interface UnlockingAccountAsset {
	readonly delegateAddress: Buffer;
	readonly amount: bigint;
	readonly unvoteHeight: number;
}

export interface UnlockingInfoJSON {
	readonly delegateAddress: string;
	readonly amount: string;
	readonly unvoteHeight: number;
	readonly minUnlockHeight: number;
}

export interface VoteAccountAsset {
	readonly delegateAddress: Buffer;
	// Amount for some delegate can be updated
	amount: bigint;
}

export interface DPOSAccountProps {
	dpos: {
		delegate: {
			username: string;
			pomHeights: number[];
			consecutiveMissedBlocks: number;
			lastForgedHeight: number;
			isBanned: boolean;
			totalVotesReceived: bigint;
		};
		sentVotes: VoteAccountAsset[];
		unlocking: UnlockingAccountAsset[];
	};
}

export interface UnlockTransactionAssetContext {
	readonly unlockObjects: ReadonlyArray<UnlockingAccountAsset>;
}

export interface RegisterTransactionAssetContext {
	readonly username: string;
}

export interface VoteTransactionAssetContext {
	readonly votes: ReadonlyArray<VoteAccountAsset>;
}

export interface BlockHeaderAssetForDPOS {
	readonly seedReveal: Buffer;
	readonly maxHeightPreviouslyForged: number;
	readonly maxHeightPrevoted: number;
}

export interface PomTransactionAssetContext {
	readonly header1: BlockHeader<BlockHeaderAssetForDPOS>;
	readonly header2: BlockHeader<BlockHeaderAssetForDPOS>;
}

export interface DelegateWeight {
	readonly address: Buffer;
	readonly voteWeight: bigint;
}

export interface VoteWeight {
	readonly round: number;
	readonly delegates: ReadonlyArray<DelegateWeight>;
}

export type VoteWeights = VoteWeight[];

export interface DecodedVoteWeights {
	voteWeights: VoteWeights;
}

type Grow<T, A extends T[]> = ((x: T, ...xs: A) => void) extends (...a: infer X) => void
	? X
	: never;
type GrowToSize<T, A extends T[], N extends number> = {
	readonly 0: A;
	readonly 1: GrowToSize<T, Grow<T, A>, N>;
}[A['length'] extends N ? 0 : 1];

export type FixedLengthArray<T, N extends number> = GrowToSize<T, [], N>;

// Look for a way to define buffer type with fixed size
export type RandomSeed = Buffer;
