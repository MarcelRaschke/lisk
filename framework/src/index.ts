/*
 * Copyright Â© 2019 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

export {
	Transaction,
	transactionSchema,
	readGenesisBlockJSON,
	blockHeaderSchema,
	blockSchema,
	signingBlockHeaderSchema,
	getAccountSchemaWithDefault,
	getGenesisBlockHeaderAssetSchema,
	blockHeaderAssetSchema,
} from '@liskhq/lisk-chain';
export {
	BaseModule,
	BaseAsset,
	TokenModule,
	TokenTransferAsset,
	SequenceModule,
	KeysModule,
	KeysRegisterAsset,
	DPoSModule,
	DPoSRegisterAsset,
	DPoSVoteAsset,
	DPoSUnlockAsset,
	DPoSPoMAsset,
} from './modules';
export { Application } from './application';
export { systemDirs } from './system_dirs';
export { BasePlugin, BlockHeaderJSON, PluginInfo, PluginCodec } from './plugins/base_plugin';
export { IPCChannel } from './controller/channels';
export type { BaseChannel } from './controller/channels';
export type { EventsDefinition, EventCallback } from './controller/event';
export type { ActionsDefinition, ActionHandler } from './controller/action';
export * as testing from './testing';
export * from './types';
