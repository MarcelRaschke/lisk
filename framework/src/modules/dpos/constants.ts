/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

// State store related constants
export const CHAIN_STATE_DELEGATE_USERNAMES = 'dpos:delegateUsernames';
export const CHAIN_STATE_DELEGATE_VOTE_WEIGHTS = 'dpos:delegateVoteWeights';

export const TEN_UNIT = BigInt(10) * BigInt(10) ** BigInt(8);
export const MAX_VOTE = 10;
export const MAX_UNLOCKING = 20;
export const MAX_INT64 = '9223372036854775807';
export const AMOUNT_MULTIPLIER_FOR_VOTES = BigInt(10) * BigInt(10) ** BigInt(8);
export const WAIT_TIME_VOTE = 2000;
export const WAIT_TIME_SELF_VOTE = 260000;
export const VOTER_PUNISH_TIME = 260000;
export const SELF_VOTE_PUNISH_TIME = 780000;
export const MAX_PUNISHABLE_BLOCK_HEIGHT_DIFFERENCE = 260000;
export const MAX_POM_HEIGHTS = 5;
export const MAX_CONSECUTIVE_MISSED_BLOCKS = 50;
export const MAX_LAST_FORGED_HEIGHT_DIFF = 260000;

export const DEFAULT_ACTIVE_DELEGATE = 101;
export const DEFAULT_STANDBY_DELEGATE = 2;
export const DEFAULT_ROUND_OFFSET = 2;

// Vote weight is capped at the self-vote * DEFAULT_VOTE_WEIGHT_CAP_RATE
export const DEFAULT_VOTE_WEIGHT_CAP_RATE = 10;
export const DEFAULT_STANDBY_THRESHOLD = BigInt(1000) * BigInt(10) ** BigInt(8);

// Punishment period is 780k block height by default
export const PUNISHMENT_PERIOD = 780000;
