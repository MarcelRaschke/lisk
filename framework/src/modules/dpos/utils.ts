/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { verifyData } from '@liskhq/lisk-cryptography';
import { Account } from '@liskhq/lisk-chain';
import { DPOSAccountProps, UnlockingAccountAsset } from './types';
import {
	PUNISHMENT_PERIOD,
	SELF_VOTE_PUNISH_TIME,
	VOTER_PUNISH_TIME,
	WAIT_TIME_SELF_VOTE,
	WAIT_TIME_VOTE,
} from './constants';

export const sortUnlocking = (unlocks: UnlockingAccountAsset[]): void => {
	unlocks.sort((a, b) => {
		if (!a.delegateAddress.equals(b.delegateAddress)) {
			return a.delegateAddress.compare(b.delegateAddress);
		}
		if (a.unvoteHeight !== b.unvoteHeight) {
			return b.unvoteHeight - a.unvoteHeight;
		}
		const diff = b.amount - a.amount;
		if (diff > BigInt(0)) {
			return 1;
		}
		if (diff < BigInt(0)) {
			return -1;
		}

		return 0;
	});
};

export const getMinPunishedHeight = (
	sender: Account<DPOSAccountProps>,
	delegate: Account<DPOSAccountProps>,
): number => {
	if (delegate.dpos.delegate.pomHeights.length === 0) {
		return 0;
	}

	const lastPomHeight = Math.max(...delegate.dpos.delegate.pomHeights);

	// https://github.com/LiskHQ/lips/blob/master/proposals/lip-0024.md#update-to-validity-of-unlock-transaction
	return sender.address.equals(delegate.address)
		? lastPomHeight + SELF_VOTE_PUNISH_TIME
		: lastPomHeight + VOTER_PUNISH_TIME;
};

export const getPunishmentPeriod = (
	sender: Account<DPOSAccountProps>,
	delegateAccount: Account<DPOSAccountProps>,
	lastBlockHeight: number,
): number => {
	const currentHeight = lastBlockHeight + 1;
	const minPunishedHeight = getMinPunishedHeight(sender, delegateAccount);
	const remainingBlocks = minPunishedHeight - currentHeight;

	return remainingBlocks < 0 ? 0 : remainingBlocks;
};

export const getMinWaitingHeight = (
	senderAddress: Buffer,
	delegateAddress: Buffer,
	unlockObject: UnlockingAccountAsset,
): number =>
	unlockObject.unvoteHeight +
	(senderAddress.equals(delegateAddress) ? WAIT_TIME_SELF_VOTE : WAIT_TIME_VOTE);

export const getWaitingPeriod = (
	senderAddress: Buffer,
	delegateAddress: Buffer,
	lastBlockHeight: number,
	unlockObject: UnlockingAccountAsset,
): number => {
	const currentHeight = lastBlockHeight + 1;
	const minWaitingHeight = getMinWaitingHeight(senderAddress, delegateAddress, unlockObject);
	const remainingBlocks = minWaitingHeight - currentHeight;

	return remainingBlocks < 0 ? 0 : remainingBlocks;
};

export const isNullCharacterIncluded = (input: string): boolean =>
	new RegExp(/\\0|\\u0000|\\x00/).test(input);

export const isUsername = (username: string): boolean => {
	if (isNullCharacterIncluded(username)) {
		return false;
	}

	if (username !== username.trim().toLowerCase()) {
		return false;
	}

	return /^[a-z0-9!@$&_.]+$/g.test(username);
};

export const validateSignature = (
	tag: string,
	networkIdentifier: Buffer,
	publicKey: Buffer,
	signature: Buffer,
	bytes: Buffer,
): boolean => verifyData(tag, networkIdentifier, bytes, signature, publicKey);

export const isCurrentlyPunished = (height: number, pomHeights: ReadonlyArray<number>): boolean => {
	if (pomHeights.length === 0) {
		return false;
	}
	const lastPomHeight = Math.max(...pomHeights);
	if (height - lastPomHeight < PUNISHMENT_PERIOD) {
		return true;
	}

	return false;
};
