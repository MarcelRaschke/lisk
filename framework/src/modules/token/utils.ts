/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
import { Block } from '@liskhq/lisk-chain';

export const getTotalFees = (
	block: Block,
	minFeePerByte: bigint,
	baseFees: ReadonlyArray<{ assetID: number; baseFee: string; moduleID: number }>,
): { readonly totalFee: bigint; readonly totalMinFee: bigint } =>
	block.payload.reduce(
		(prev, current) => {
			const baseFee =
				baseFees.find(
					(fee: { moduleID: number; assetID: number }) =>
						fee.moduleID === current.moduleID && fee.assetID === current.assetID,
				)?.baseFee ?? BigInt(0);
			const minFee = minFeePerByte * BigInt(current.getBytes().length) + BigInt(baseFee);

			return {
				totalFee: prev.totalFee + current.fee,
				totalMinFee: prev.totalMinFee + minFee,
			};
		},
		{ totalFee: BigInt(0), totalMinFee: BigInt(0) },
	);
