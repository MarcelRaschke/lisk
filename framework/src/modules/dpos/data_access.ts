/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { codec } from '@liskhq/lisk-codec';
import { StateStore } from '../../types';
import {
	DecodedVoteWeights,
	DelegatePersistedUsernames,
	RegisteredDelegates,
	VoteWeights,
} from './types';
import { CHAIN_STATE_DELEGATE_USERNAMES, CHAIN_STATE_DELEGATE_VOTE_WEIGHTS } from './constants';
import { delegatesUserNamesSchema, voteWeightsSchema } from './schema';

export const getRegisteredDelegates = async (
	store: StateStore,
): Promise<DelegatePersistedUsernames> => {
	const usernamesBuffer = await store.chain.get(CHAIN_STATE_DELEGATE_USERNAMES);
	if (!usernamesBuffer) {
		return { registeredDelegates: [] };
	}
	const parsedUsernames = codec.decode<RegisteredDelegates>(
		delegatesUserNamesSchema,
		usernamesBuffer,
	);

	return parsedUsernames;
};

export const setRegisteredDelegates = async (
	store: StateStore,
	usernames: DelegatePersistedUsernames,
): Promise<void> => {
	usernames.registeredDelegates.sort((a, b) => a.address.compare(b.address));

	await store.chain.set(
		CHAIN_STATE_DELEGATE_USERNAMES,
		codec.encode(delegatesUserNamesSchema, usernames),
	);
};

export const getVoteWeights = async (stateStore: StateStore): Promise<VoteWeights> => {
	const voteWeights = await stateStore.chain.get(CHAIN_STATE_DELEGATE_VOTE_WEIGHTS);
	if (!voteWeights) {
		return [];
	}

	const voteWeightsDecoded = codec.decode<DecodedVoteWeights>(voteWeightsSchema, voteWeights);
	return voteWeightsDecoded.voteWeights;
};

export const setVoteWeights = async (
	stateStore: StateStore,
	voteWeights: VoteWeights,
): Promise<void> => {
	await stateStore.chain.set(
		CHAIN_STATE_DELEGATE_VOTE_WEIGHTS,
		codec.encode(voteWeightsSchema, { voteWeights }),
	);
};

export const deleteVoteWeightsUntilRound = async (
	round: number,
	stateStore: StateStore,
): Promise<void> => {
	const voteWeights = await getVoteWeights(stateStore);
	const newVoteWeights = voteWeights.filter(vw => vw.round >= round);
	await setVoteWeights(stateStore, newVoteWeights);
};
