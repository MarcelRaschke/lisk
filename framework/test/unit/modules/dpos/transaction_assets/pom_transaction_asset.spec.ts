/*
 * Copyright Â© 2020 Lisk Foundation
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Unless otherwise agreed in a custom licensing agreement with the Lisk Foundation,
 * no part of this software, including this file, may be copied, modified,
 * propagated, or distributed except according to the terms contained in the
 * LICENSE file.
 *
 * Removal or modification of this copyright notice is prohibited.
 */

import { when } from 'jest-when';
import { BlockHeader, Account } from '@liskhq/lisk-chain';
import { objects } from '@liskhq/lisk-utils';
import { getAddressAndPublicKeyFromPassphrase, getRandomBytes } from '@liskhq/lisk-cryptography';
import { ApplyAssetContext, ValidateAssetContext } from '../../../../../src';
import {
	BlockHeaderAssetForDPOS,
	DPOSAccountProps,
	PomTransactionAssetContext,
	DPoSModule,
} from '../../../../../src/modules/dpos';
import { PomTransactionAsset } from '../../../../../src/modules/dpos/transaction_assets/pom_transaction_asset';
import * as dposUtils from '../../../../../src/modules/dpos/utils';
import { liskToBeddows } from '../../../../utils/assets';
import * as testing from '../../../../../src/testing';

const { StateStoreMock } = testing.mocks;

describe('PomTransactionAsset', () => {
	const lastBlockHeight = 8760000;
	const lastBlockReward = liskToBeddows(69);
	let transactionAsset: PomTransactionAsset;
	let applyContext: ApplyAssetContext<PomTransactionAssetContext>;
	let validateContext: ValidateAssetContext<PomTransactionAssetContext>;
	let sender: any;
	let stateStoreMock: testing.mocks.StateStoreMock;
	let misBehavingDelegate: Account<DPOSAccountProps>;
	let normalDelegate: Account<DPOSAccountProps>;
	let header1: BlockHeader<BlockHeaderAssetForDPOS>;
	let header2: BlockHeader<BlockHeaderAssetForDPOS>;

	beforeEach(() => {
		sender = testing.fixtures.createDefaultAccount<DPOSAccountProps>([DPoSModule], {});

		const {
			address: delegate1Address,
			publicKey: delegate1PublicKey,
		} = getAddressAndPublicKeyFromPassphrase(getRandomBytes(20).toString('utf8'));

		misBehavingDelegate = testing.fixtures.createDefaultAccount<DPOSAccountProps>([DPoSModule], {
			address: delegate1Address,
			dpos: { delegate: { username: 'misBehavingDelegate' } },
		});
		normalDelegate = testing.fixtures.createDefaultAccount<DPOSAccountProps>([DPoSModule], {
			dpos: { delegate: { username: 'normalDelegate' } },
		});
		const { id: id1, ...fakeBlockHeader1 } = testing.createFakeBlockHeader({
			generatorPublicKey: delegate1PublicKey,
		});
		const { id: id2, ...fakeBlockHeader2 } = testing.createFakeBlockHeader({
			generatorPublicKey: delegate1PublicKey,
		});

		header1 = fakeBlockHeader1 as BlockHeader;
		header2 = fakeBlockHeader2 as BlockHeader;

		stateStoreMock = new StateStoreMock({
			accounts: objects.cloneDeep([sender, misBehavingDelegate, normalDelegate]),
			lastBlockHeaders: [{ height: lastBlockHeight }] as any,
			lastBlockReward,
		});

		transactionAsset = new PomTransactionAsset();

		const transaction = {
			senderAddress: sender.address,
		} as any;

		const asset = {
			header1: {} as any,
			header2: {} as any,
		};

		applyContext = testing.createApplyAssetContext<PomTransactionAssetContext>({
			transaction,
			asset,
			stateStore: stateStoreMock,
			reducerHandler: {
				invoke: jest.fn(),
			},
		});

		validateContext = testing.createValidateAssetContext({ asset, transaction });

		jest.spyOn(stateStoreMock.account, 'get');
		jest.spyOn(stateStoreMock.account, 'set');
	});

	describe('constructor', () => {
		it('should have valid id', () => {
			expect(transactionAsset.id).toEqual(3);
		});

		it('should have valid name', () => {
			expect(transactionAsset.name).toEqual('reportDelegateMisbehavior');
		});

		it('should have valid schema', () => {
			expect(transactionAsset.schema).toMatchSnapshot();
		});
	});

	describe('validate', () => {
		it('should throw error when generatorPublicKey does not match', () => {
			validateContext.asset = {
				header1: {
					...header1,
				},
				header2: { ...header2, generatorPublicKey: getRandomBytes(20) },
			};

			expect(() => transactionAsset.validate(validateContext)).toThrow(
				'BlockHeaders are not contradicting as per BFT violation rules.',
			);
		});

		it('should throw error when both headers are identical', () => {
			validateContext.asset = {
				header1: {
					...header1,
				},
				header2: { ...header1 },
			};

			expect(() => transactionAsset.validate(validateContext)).toThrow(
				'BlockHeaders are not contradicting as per BFT violation rules.',
			);
		});

		it('should not throw error when first height is equal to second height but equal maxHeightPrevoted', () => {
			validateContext.asset = {
				header1: {
					...header1,
					height: 10999,
					asset: { ...header1.asset, maxHeightPrevoted: 1099 },
				},
				header2: { ...header2, height: 10999 },
			};

			expect(() => transactionAsset.validate(validateContext)).not.toThrow();
		});

		it('should not throw error when first height is greater than the second height but equal maxHeightPrevoted', () => {
			validateContext.asset = {
				header1: {
					...header1,
					height: 10999,
					asset: { ...header1.asset, maxHeightPrevoted: 1099 },
				},
				header2: { ...header2, height: 11999 },
			};

			expect(() => transactionAsset.validate(validateContext)).not.toThrow();
		});

		it("should not throw error when height is greater than the second header's maxHeightPreviouslyForged", () => {
			validateContext.asset = {
				header1: {
					...header1,
					height: 120,
				},
				header2: {
					...header2,
					height: 123,
					asset: { ...header1.asset, maxHeightPreviouslyForged: 98 },
				},
			};

			expect(() => transactionAsset.validate(validateContext)).not.toThrow();
		});

		it('should not throw error when maxHeightPrevoted is greater than the second maxHeightPrevoted', () => {
			validateContext.asset = {
				header1: {
					...header1,
					height: 133,
					asset: { ...header1.asset, maxHeightPrevoted: 101 },
				},
				header2: { ...header2, height: 123, asset: { ...header1.asset, maxHeightPrevoted: 98 } },
			};

			expect(() => transactionAsset.validate(validateContext)).not.toThrow();
		});

		it('should throw error when headers are not contradicting', () => {
			validateContext.asset = {
				header1: {
					...header1,
				},
				header2: { ...header1 },
			};

			expect(() => transactionAsset.validate(validateContext)).toThrow(
				'BlockHeaders are not contradicting as per BFT violation rules.',
			);
		});
	});

	describe('apply', () => {
		const block1Height = lastBlockHeight - 768;
		const block2Height = block1Height + 15;

		beforeEach(() => {
			jest.spyOn(dposUtils, 'validateSignature').mockReturnValue(true);

			applyContext.asset = {
				header1: { ...header1, height: block1Height },
				header2: { ...header2, height: block2Height },
			};
		});

		afterEach(() => {
			(dposUtils.validateSignature as any).mockClear();
		});

		it('should not throw error with valid transactions', async () => {
			await expect(transactionAsset.apply(applyContext)).resolves.toBeUndefined();
		});

		it('should throw error if |header1.height - h| >= 260000', async () => {
			applyContext.asset = {
				...applyContext.asset,
				header1: {
					...applyContext.asset.header1,
					height: lastBlockHeight - 260000,
				},
			};

			await expect(transactionAsset.apply(applyContext)).rejects.toThrow(
				'Difference between header1.height and current height must be less than 260000.',
			);
		});

		it('should throw error if |header2.height - h| >= 260000', async () => {
			applyContext.asset = {
				...applyContext.asset,
				header2: {
					...applyContext.asset.header2,
					height: lastBlockHeight - 260000,
				},
			};

			await expect(transactionAsset.apply(applyContext)).rejects.toThrow(
				'Difference between header2.height and current height must be less than 260000.',
			);
		});

		it('should throw error when header1 is not properly signed', async () => {
			when(dposUtils.validateSignature as any)
				.calledWith(
					applyContext.asset.header1.generatorPublicKey,
					applyContext.asset.header1.signature,
					expect.any(Buffer),
				)
				.mockReturnValue(false);
			when(dposUtils.validateSignature as any)
				.calledWith(
					applyContext.asset.header2.generatorPublicKey,
					applyContext.asset.header2.signature,
					expect.any(Buffer),
				)
				.mockReturnValue(true);

			await expect(transactionAsset.apply(applyContext)).rejects.toThrow(
				'Invalid block signature for header 1',
			);
		});

		it('should throw error when header2 is not properly signed', async () => {
			when(dposUtils.validateSignature as any)
				.calledWith(
					'LSK_BH_',
					expect.any(Buffer),
					applyContext.asset.header1.generatorPublicKey,
					applyContext.asset.header1.signature,
					expect.any(Buffer),
				)
				.mockReturnValue(true);
			when(dposUtils.validateSignature as any)
				.calledWith(
					'LSK_BH_',
					expect.any(Buffer),
					applyContext.asset.header2.generatorPublicKey,
					applyContext.asset.header2.signature,
					expect.any(Buffer),
				)
				.mockReturnValue(false);

			await expect(transactionAsset.apply(applyContext)).rejects.toThrow(
				'Invalid block signature for header 2',
			);
		});

		it('should throw error if misbehaving account is not a delegate', async () => {
			const updatedDelegateAccount = objects.cloneDeep(misBehavingDelegate);
			updatedDelegateAccount.dpos.delegate.username = '';
			await stateStoreMock.account.set(misBehavingDelegate.address, updatedDelegateAccount);

			await expect(transactionAsset.apply(applyContext)).rejects.toThrow(
				'Account is not a delegate',
			);
		});

		it('should throw error if misbehaving account is already banned', async () => {
			const updatedDelegateAccount = objects.cloneDeep(misBehavingDelegate);
			updatedDelegateAccount.dpos.delegate.isBanned = true;
			await stateStoreMock.account.set(misBehavingDelegate.address, updatedDelegateAccount);

			await expect(transactionAsset.apply(applyContext)).rejects.toThrow(
				'Cannot apply proof-of-misbehavior. Delegate is already banned.',
			);
		});

		it('should throw error if misbehaving account is already punished at height h', async () => {
			const updatedDelegateAccount = objects.cloneDeep(misBehavingDelegate);
			updatedDelegateAccount.dpos.delegate.pomHeights = [applyContext.asset.header1.height + 10];
			await stateStoreMock.account.set(misBehavingDelegate.address, updatedDelegateAccount);

			await expect(transactionAsset.apply(applyContext)).rejects.toThrow(
				'Cannot apply proof-of-misbehavior. Delegate is already punished.',
			);
		});

		it('should reward the sender with last block reward if delegate have enough balance', async () => {
			const remainingBalance = lastBlockReward + BigInt('10000000000');
			const minRemainingBalance = BigInt('5000000');

			when(applyContext.reducerHandler.invoke as any)
				.calledWith('token:getBalance', { address: misBehavingDelegate.address })
				.mockResolvedValue(remainingBalance as never);
			when(applyContext.reducerHandler.invoke as any)
				.calledWith('token:getMinRemainingBalance')
				.mockResolvedValue(minRemainingBalance as never);

			await transactionAsset.apply(applyContext);

			expect(applyContext.reducerHandler.invoke).toHaveBeenCalledWith('token:credit', {
				address: applyContext.transaction.senderAddress,
				amount: lastBlockReward,
			});
		});

		it('should not reward the sender if delegate does not have enough minimum remaining balance', async () => {
			const remainingBalance = BigInt(100);
			const minRemainingBalance = BigInt('5000000');

			when(applyContext.reducerHandler.invoke as any)
				.calledWith('token:getBalance', { address: misBehavingDelegate.address })
				.mockResolvedValue(remainingBalance as never);
			when(applyContext.reducerHandler.invoke as any)
				.calledWith('token:getMinRemainingBalance')
				.mockResolvedValue(minRemainingBalance as never);

			await transactionAsset.apply(applyContext);

			// If amount is zero, it should not call the credit
			expect(applyContext.reducerHandler.invoke).not.toHaveBeenCalledWith('token:credit', {
				address: applyContext.transaction.senderAddress,
				amount: BigInt(0),
			});
			expect(applyContext.reducerHandler.invoke).not.toHaveBeenCalledWith('token:debit', {
				address: applyContext.transaction.senderAddress,
				amount: BigInt(0),
			});
		});

		it('should add (remaining balance - min remaining balance) of delegate to balance of the sender if delegate balance is less than last block reward', async () => {
			const remainingBalance = lastBlockReward - BigInt(1);
			const minRemainingBalance = BigInt('5000000');

			when(applyContext.reducerHandler.invoke as any)
				.calledWith('token:getBalance', { address: misBehavingDelegate.address })
				.mockResolvedValue(remainingBalance as never);
			when(applyContext.reducerHandler.invoke as any)
				.calledWith('token:getMinRemainingBalance')
				.mockResolvedValue(minRemainingBalance as never);

			await transactionAsset.apply(applyContext);

			expect(applyContext.reducerHandler.invoke).toHaveBeenCalledWith('token:credit', {
				address: applyContext.transaction.senderAddress,
				amount: remainingBalance - minRemainingBalance,
			});
		});

		it('should append height h to pomHeights property of misbehaving account', async () => {
			await transactionAsset.apply(applyContext);

			const updatedDelegate = await stateStoreMock.account.get<Account<DPOSAccountProps>>(
				misBehavingDelegate.address,
			);

			expect(updatedDelegate.dpos.delegate.pomHeights).toEqual([lastBlockHeight + 1]);
		});

		it('should set isBanned property to true is pomHeights.length === 5', async () => {
			const pomHeights = [500, 1000, 2000, 4550];
			const updatedDelegateAccount = objects.cloneDeep(misBehavingDelegate);
			updatedDelegateAccount.dpos.delegate.pomHeights = objects.cloneDeep(pomHeights);
			updatedDelegateAccount.dpos.delegate.isBanned = false;
			await stateStoreMock.account.set(misBehavingDelegate.address, updatedDelegateAccount);

			await transactionAsset.apply(applyContext);

			const updatedDelegate = await stateStoreMock.account.get<Account<DPOSAccountProps>>(
				misBehavingDelegate.address,
			);

			expect(updatedDelegate.dpos.delegate.pomHeights).toEqual([
				...pomHeights,
				lastBlockHeight + 1,
			]);
			expect(updatedDelegate.dpos.delegate.pomHeights).toHaveLength(5);
			expect(updatedDelegate.dpos.delegate.isBanned).toBeTrue();
		});

		it('should not return balance if sender and delegate account are same', async () => {
			(applyContext.transaction as any).senderAddress = misBehavingDelegate.address;

			await expect(transactionAsset.apply(applyContext)).resolves.toBeUndefined();
		});
	});
});
