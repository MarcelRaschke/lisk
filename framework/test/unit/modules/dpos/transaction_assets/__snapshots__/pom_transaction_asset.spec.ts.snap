// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PomTransactionAsset constructor should have valid schema 1`] = `
Object {
  "$id": "lisk/dpos/pom",
  "properties": Object {
    "header1": Object {
      "$id": "block-header1",
      "fieldNumber": 1,
      "properties": Object {
        "asset": Object {
          "fieldNumber": 8,
          "properties": Object {
            "maxHeightPreviouslyForged": Object {
              "dataType": "uint32",
              "fieldNumber": 1,
            },
            "maxHeightPrevoted": Object {
              "dataType": "uint32",
              "fieldNumber": 2,
            },
            "seedReveal": Object {
              "dataType": "bytes",
              "fieldNumber": 3,
            },
          },
          "required": Array [
            "maxHeightPreviouslyForged",
            "maxHeightPrevoted",
            "seedReveal",
          ],
          "type": "object",
        },
        "generatorPublicKey": Object {
          "dataType": "bytes",
          "fieldNumber": 6,
        },
        "height": Object {
          "dataType": "uint32",
          "fieldNumber": 3,
        },
        "previousBlockID": Object {
          "dataType": "bytes",
          "fieldNumber": 4,
        },
        "reward": Object {
          "dataType": "uint64",
          "fieldNumber": 7,
        },
        "signature": Object {
          "dataType": "bytes",
          "fieldNumber": 9,
        },
        "timestamp": Object {
          "dataType": "uint32",
          "fieldNumber": 2,
        },
        "transactionRoot": Object {
          "dataType": "bytes",
          "fieldNumber": 5,
        },
        "version": Object {
          "dataType": "uint32",
          "fieldNumber": 1,
        },
      },
      "required": Array [
        "version",
        "timestamp",
        "height",
        "previousBlockID",
        "transactionRoot",
        "generatorPublicKey",
        "reward",
        "asset",
      ],
      "type": "object",
    },
    "header2": Object {
      "$id": "block-header2",
      "fieldNumber": 2,
      "properties": Object {
        "asset": Object {
          "fieldNumber": 8,
          "properties": Object {
            "maxHeightPreviouslyForged": Object {
              "dataType": "uint32",
              "fieldNumber": 1,
            },
            "maxHeightPrevoted": Object {
              "dataType": "uint32",
              "fieldNumber": 2,
            },
            "seedReveal": Object {
              "dataType": "bytes",
              "fieldNumber": 3,
            },
          },
          "required": Array [
            "maxHeightPreviouslyForged",
            "maxHeightPrevoted",
            "seedReveal",
          ],
          "type": "object",
        },
        "generatorPublicKey": Object {
          "dataType": "bytes",
          "fieldNumber": 6,
        },
        "height": Object {
          "dataType": "uint32",
          "fieldNumber": 3,
        },
        "previousBlockID": Object {
          "dataType": "bytes",
          "fieldNumber": 4,
        },
        "reward": Object {
          "dataType": "uint64",
          "fieldNumber": 7,
        },
        "signature": Object {
          "dataType": "bytes",
          "fieldNumber": 9,
        },
        "timestamp": Object {
          "dataType": "uint32",
          "fieldNumber": 2,
        },
        "transactionRoot": Object {
          "dataType": "bytes",
          "fieldNumber": 5,
        },
        "version": Object {
          "dataType": "uint32",
          "fieldNumber": 1,
        },
      },
      "required": Array [
        "version",
        "timestamp",
        "height",
        "previousBlockID",
        "transactionRoot",
        "generatorPublicKey",
        "reward",
        "asset",
      ],
      "type": "object",
    },
  },
  "required": Array [
    "header1",
    "header2",
  ],
  "type": "object",
}
`;
